Kernel/FileSystem/ISO9660FileSystem.cpp

```diff
@@ -0,0 +1,726 @@
+            m_primary_volume = adopt_own_if_nonnull(new ISO::PrimaryVolumeDescriptor(*primary_volume));
+            break;
+        }
+        case ISO::VolumeDescriptorType::BootRecord:
+        case ISO::VolumeDescriptorType::SupplementaryOrEnhancedVolumeDescriptor:
+        case ISO::VolumeDescriptorType::VolumePartitionDescriptor: {
+            break;
+        }
+        case ISO::VolumeDescriptorType::VolumeDescriptorSetTerminator: {
+            goto all_headers_read;
+        }
+        default:
+            dbgln_if(ISO9660_DEBUG, "Unexpected volume descriptor type {} in volume set", static_cast<u8>(header->type));
+            return false;
+        }
+
+        current_block_index++;
+    }
+
+all_headers_read:
+    if (!m_primary_volume) {
+        dbgln_if(ISO9660_DEBUG, "Could not find primary volume");
+        return false;
+    }
+
+    m_logical_block_size = LittleEndian { m_primary_volume->logical_block_size.little };
+    return true;
+}
+
+bool ISO9660FS::create_root_inode()
+{
+    if (!m_primary_volume) {
+        dbgln_if(ISO9660_DEBUG, "Primary volume doesn't exist, can't create root inode");
+        return false;
+    }
+
+    auto maybe_inode = ISO9660Inode::create_from_directory_record(*this, &m_primary_volume->root_directory_record_header, {});
+    if (maybe_inode.is_error()) {
+        // FIXME: Fix the FileSystem::initialize contract to return a KResult.
+        dbgln("Failed to allocate root node for the ISO 9660 filesystem!");
+        return false;
+    }
+
+    m_root_inode = maybe_inode.release_value();
+    return true;
+}
+
+bool ISO9660FS::calculate_inode_count() const
+{
+    if (!m_primary_volume) {
+        dbgln_if(ISO9660_DEBUG, "Primary volume doesn't exist, can't calculate inode count");
+        return false;
+    }
+
+    m_cached_inode_count = 1;
+
+    auto traversal_result = visit_directory_record(&m_primary_volume->root_directory_record_header, [&](ISO::DirectoryRecordHeader const* header) {
+        if (header == nullptr) {
+            return RecursionDecision::Continue;
+        }
+
+        m_cached_inode_count += 1;
+
+        if (has_flag(header->file_flags, ISO::FileFlags::Directory)) {
+            if (header->file_identifier_length == 1) {
+                auto file_identifier = reinterpret_cast<u8 const*>(header + 1);
+                if (file_identifier[0] == '\0' || file_identifier[0] == '\1') {
+                    return RecursionDecision::Continue;
+                }
+            }
+
+            return RecursionDecision::Recurse;
+        }
+
+        return RecursionDecision::Continue;
+    });
+
+    if (traversal_result.is_error()) {
+        dbgln_if(ISO9660_DEBUG, "Failed to traverse for caching inode count: {}", traversal_result);
+        return false;
+    }
+
+    return true;
+}
+
+KResult ISO9660FS::visit_directory_record(ISO::DirectoryRecordHeader const* record, Function<KResultOr<RecursionDecision>(ISO::DirectoryRecordHeader const*)> const& visitor) const
```

Comment:
This function does not support `record` being null, so it should take `record` as a reference.